### PR TITLE
Allow tuning loader options

### PR DESCRIPTION
**Pull request by lucaviolanti**
This addresses [#332](https://github.com/circe/circe-yaml/issues/332).

SnakeYaml sets various default limits, and supports overriding these values via [LoaderOptions](https://www.javadoc.io/doc/org.yaml/snakeyaml/latest/org/yaml/snakeyaml/LoaderOptions.html).

This PR enables customisation of these limits, following a similar pattern as https://github.com/circe/circe-yaml/pull/205.

### CODE REVIEW

File: src/main/scala/io/circe/yaml/Parser.scala [hunk 1]
```diff
@@ -1,22 +1,26 @@
...
 import org.yaml.snakeyaml.constructor.SafeConstructor
 import org.yaml.snakeyaml.nodes._
 import scala.collection.JavaConverters._
 
+
 final case class Parser(
-  maxAliasesForCollections: Int = 50
+  maxAliasesForCollections: Int = 50,
+  nestingDepthLimit: Int = 50,
+  codePointLimit: Int = 3 * 1024 * 1024 // 3 MB
```
Comment (zarthross) [inline src/main/scala/io/circe/yaml/Parser.scala:16]: Should we add all the other remaining `LoaderOptions` as well while we are at it?

Also, this is certainly going to break bin compat with 0.14.2.  Can you pull latest from master and run `mimaReportBinaryIssues`?

**Comment by zarthross** [comment]
Thanks for the PR!

File: src/main/scala/io/circe/yaml/Parser.scala (hunk 1 above)
Comment (lucaviolanti) [reply src/main/scala/io/circe/yaml/Parser.scala]: I've addressed all bin compat issues, seems OK to me now.

With respect to the other (boolean) `LoaderOptions`, a lengthy investigation we had on this revealed that they are ignored when using the underlying `Yaml.compose` method in snake-yaml - as circe-yaml does. They only affect the behaviour of snake-yaml's `Yaml.loadAs`, which is currently not used by circe-yaml (imo for the best).

File: src/main/scala/io/circe/yaml/Parser.scala [hunk 2]
```diff
@@ -40,19 +43,43 @@ final case class Parser(
...
   private[yaml] object CustomTag {
     def unapply(tag: Tag): Option[String] = if (!tag.startsWith(Tag.PREFIX))
       Some(tag.getValue)
     else
       None
   }
 
-  private[yaml] class FlatteningConstructor extends SafeConstructor {
+  private[yaml] class FlatteningConstructor(val loaderOptions: LoaderOptions = new LoaderOptions)
+      extends SafeConstructor(loaderOptions) {
```
[the remark this answers is not in the document]
Comment (lucaviolanti) [reply src/main/scala/io/circe/yaml/Parser.scala]: Should be fixed now :)

**Comment by zarthross** [comment]
```
[error] circe-yaml: Failed binary compatibility check against io.circe:circe-yaml_3:0.14.2! Found 1 potential problems
[error]  * static method yamlToJson(org.yaml.snakeyaml.nodes.Node)scala.util.Either in class io.circe.yaml.Parser does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("io.circe.yaml.Parser.yamlToJson")
[error] stack trace is suppressed; run last mimaReportBinaryIssues for the full output
[error] (mimaReportBinaryIssues) Failed binary compatibility check against io.circe:circe-yaml_3:0.14.2! Found 1 potential problems
[error] Total time: 9 s, completed Nov 18, 2022 4:06:24 PM
```

**Review by zarthross — APPROVED** [review]

**Review by zarthross — APPROVED** [review]

**Pull request merged**